Some parallelization of importer input steps

tinwelint opened the pull request:
There's plumbing inside the batch importer to communicating with and requesting more threads from input steps. Then the focus of this change set is to parallelize data generating input, which is also made much simpler. Reading cached input can also be parallelized now.

This paves the way for adding parallelized reading/parsing of CSV data and other inputs.

tinwelint committed:
InputIterator implements Parallelizable

as to be able to listen to requests from the batch importer
about the need for more threads in input steps.

tinwelint committed:
QuickImport generates input data in multiple threads

also simplified the way data can be generated this way be a great deal
so that the only thing required is to implement:

- one Function<Range,InputNode[]>
- one Function<Range,InputRelationship[]>

where Range is a start id and number of entities, and the rest will
be taken care of.

Introducing TicketedProcessing which simplifies writing of parallelization
like this.

tinwelint committed:
Parallel reading of cached input data

For this the format needed to change a bit so that entities are written
in batches with a small header saying how big a batch is, byte-wise as
well as entity count-wise. This favors a setup where there's a dedicated
reader and multiple processors of that data.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/InputIterator.java:
```diff
@@ -103,6 +103,24 @@ public long position()
+        public boolean decrementNumberOfProcessors()
+        {
+            return actual.decrementNumberOfProcessors();
+        }
```
This does not delegate `setNumberOfProcessors()`

tinwelint replied:
You're absolutely right. I wonder... since there are many doing this delegation, could this be done differently, you think? Or perhaps those 4 methods could be reduced to one method:

`int changeNumberOfProcessors(int delta)`

where the returned value would be how many processors are assigned when the call exits.

chrisvest replied:
That sounds like a good idea. Same range of behavioural freedom, but a smaller interface.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/Parallelizable.java:
```diff
@@ -26,21 +26,80 @@
+        public boolean decrementNumberOfProcessors()
+        {
+            return actual.decrementNumberOfProcessors();
+        }
```
This should also delegate `setNumberOfProcessors`. It's up to `actual` to decide if it wants to use the default implementation or not.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/IteratorBatcherStep.java:
```diff
@@ -68,4 +68,23 @@ protected long position()
+    public boolean decrementNumberOfProcessors()
+    {
+        return data.decrementNumberOfProcessors();
+    }
```
This should also delegate `setNumberOfProcessors` to `data`.

tinwelint committed:
Reduces Parallelizable interface to one method

because most of the time that interface is about delegation and delegating
4 methods in loads of places turns out to be really annoying.
The interface is reduced to one method `processers(delta)` which can
be used to do everything those 4 methods could previous AND delegation
becomes very simple.

chrisvest commented on community/import-tool/src/test/java/org/neo4j/tooling/EntityDataGenerator.java:
```diff
@@ -0,0 +1,109 @@
+    private final TicketedProcessing<Range,Void,T[]> processing;
+
+    private long cursor;
+    private long position;
```
`position` is always 0. Is it referring to the character position within a hypothetical CSV file? You could make the `position()` method just return 0.

tinwelint replied:
Yeah, it's to drive the e.g. `67 MiB/s` counter in the input step printout. Whatever we'd come up with for this one it would be contrived. Would you think that simply keeping it as 0 always would be a good thing or could we think of something that may be useful?

chrisvest replied:
Okay. This is only for generating random data for testing, so I don't think it's worth the time to come up with some heuristic or contrived value. 0 will be fine, I think.

chrisvest commented on community/import-tool/src/test/java/org/neo4j/tooling/QuickImport.java:
```diff
@@ -98,12 +94,17 @@ public int denseNodeThreshold()
-                silentBadCollector( 0 ));
+
+        SimpleDataGenerator generator = new SimpleDataGenerator( nodeHeader, relationshipHeader, relationshipCount,
+                nodeCount, labelCount, relationshipTypeCount, idType );
```
The `relationshipCount` argument is slotted into the `randomSeed` parameter, which doesn't sound like it does the same thing.

tinwelint replied:
Oops, fixed it and also made it configurable w/ `--random-seed` defaulting to `currentTimeMillis()`

chrisvest commented on community/import-tool/src/test/java/org/neo4j/tooling/SimpleDataGenerator.java:
```diff
@@ -0,0 +1,87 @@
+        return batch -> new SimpleDataGeneratorBatch<>( relationshipHeader, batch.getStart(),
+                randomSeed + batch.getStart(), nodeCount, labels, relationshipTypes,
+                new InputRelationshipDeserialization( SimpleDataGenerator.this, relationshipHeader, groups ),
+                new InputRelationship[batch.getSize()] ).get();
```
You are generating a `nodeCount` number of relationships here.

tinwelint replied:
It's actually intended, on this level there's no concept of `relationshipCount` - this thing keeps generating stuff as long as there's call for it. The `nodeCount` is for knowing which bounds the generated node ids should be between.

chrisvest replied:
Okay. That's a little confusing.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/executor/DynamicTaskExecutor.java:
```diff
@@ -37,14 +37,15 @@
 public class DynamicTaskExecutor<LOCAL> implements TaskExecutor<LOCAL>
 {
-    public static final ParkStrategy DEFAULT_PARK_STRATEGY = new ParkStrategy.Park( 10 );
+    public static final ParkStrategy DEFAULT_PARK_STRATEGY = new ParkStrategy.Park( 100, NANOSECONDS );
```
Reducing the default park time from 10 milliseconds to 100 nanoseconds? No OS can context switch that quickly. You might as well be spinning or calling `Thread.yield()`. In Java 9 there will be an `Thread.onSpinWait()`.

tinwelint replied:
Yeah, perhaps unnecessary

tinwelint committed:
QuickImport accepts "random-seed" config

and fixes a mistake where wrong randomSeed was used

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/TicketedProcessing.java:
```diff
@@ -0,0 +1,185 @@
+                return next;
+            }
+            park.park( Thread.currentThread() );
+            healthCheck.run();
```
Why not use the blocking properties of your `processed` queue here, and in `submit()`?

tinwelint replied:
That would be to do the health check continously. Perhaps it's not really worth it and the health checking may be achievable in another way.

EDIT: oh it's also to check done and all that continously. Although could use `poll(time,unit)` instead of the parking. In `submit` it's a little different since we're waiting to keep ticket order as well.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/util/collection/ContinuableArrayCursorTest.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Test
+    public void shouldWorkSupplierWithEmptyArray() throws Exception
```
Your test names don't say what it means for the `ContiuableArrayCursor` to "work".

tinwelint commented:
@chrisvest most of your comments addressed I believe
